### DISCUSSION TITLE
'brew install openjdk' install jdk 17 but shows version as 16. How to install openjdk 16?

### DISCUSSION THREAD

**Discussion by thesurenk** [Homebrew/discussions#1594] [Everyday usage]
Wanted to install openjdk 16 on macOS (macOS Big Sur 11.0.1; Apple M1). `brew info openjdk` shows as 16.0.1 but on install actual install **version 17**. How can I install **openjdk 16** ?

Steps:
```
On a Apple machine with M1 chip:
$brew update
$brew install openjdk
```

But `$java -version` shows as 17

```
$ java -version
  openjdk version "17" 2021-09-14
  OpenJDK Runtime Environment Homebrew (build 17+0)
  OpenJDK 64-Bit Server VM Homebrew (build 17+0, mixed mode)

OR

$ /usr/libexec/java_home -V
Matching Java Virtual Machines (1):
  17 (arm64) "Homebrew" - "OpenJDK 17" 
       /opt/homebrew/Cellar/openjdk/16.0.1/libexec/openjdk.jdk/Contents/Home
          /opt/homebrew/Cellar/openjdk/16.0.1/libexec/openjdk.jdk/Contents/Home
```

Part of the code for the formula `openjdk` at https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/openjdk.rb is actually installing 17 but showing the version as 16.0.1 on purpose? 

```
if Hardware::CPU.arm?   
   # Temporarily use a openjdk 17 preview on Apple Silicon   
   # (because it is better than nothing)    
   url "https://github.com/openjdk/jdk/archive/refs/tags/jdk-17+24.tar.gz"    
   sha256 "9d1ea3fc63ce860e55a9be77f670b18fa7b7e5c9773dca3c70042403e1ee285c"   
   version "16.0.1"
```

**Comment by carlocab** [comment]
I'm not sure you can. OpenJDK 16 didn't compile on ARM. See https://github.com/Homebrew/homebrew-core/pull/72535. (This is why the formula installs OpenJDK 17, btw.) However, if you find a way to build it, do let us know; thanks.

**Reply by thesurenk** [reply]
Thanks @carlocab . Not an openjdk developer so I'm not sure how / what needs to happen. I'm a java application developer and will need specific versions for my work. 

If openjdk 16 does not compile on ARM, that's fine. Update the "version" information to be what is being installed and mention 16 isn't supported yet?

I also understand openjdk/java builds are releasing at a fast pace and will be hard to catch up.

**Reply by carlocab** [reply]
Does `brew info openjdk` show any caveats for you? If not, does it show any after a `brew update`?

**Comment by thesurenk** [comment]
Yes it does

```
$brew info openjdk
openjdk: stable 16.0.1 (bottled) [keg-only]
Development kit for the Java programming language
https://openjdk.java.net/
/opt/homebrew/Cellar/openjdk/16.0.1 (618 files, 281.4MB)
  Poured from bottle on 2021-05-31 at 23:05:51
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/openjdk.rb
License: GPL-2.0-only with Classpath-exception-2.0
==> Dependencies
Build: autoconf ✘
==> Requirements
Build: Xcode ✘
==> Caveats
For the system Java wrappers to find this JDK, symlink it with
  sudo ln -sfn /opt/homebrew/opt/openjdk/libexec/openjdk.jdk /Library/Java/JavaVirtualMachines/openjdk.jdk
This is a beta version of openjdk for Apple Silicon
(openjdk 17 preview).

openjdk is keg-only, which means it was not symlinked into /opt/homebrew,
because macOS provides similar software and installing this software in
parallel can cause all kinds of trouble.

If you need to have openjdk first in your PATH, run:
  echo 'export PATH="/opt/homebrew/opt/openjdk/bin:$PATH"' >> /Users/vassu/.bash_profile

For compilers to find openjdk you may need to set:
  export CPPFLAGS="-I/opt/homebrew/opt/openjdk/include"

==> Analytics
install: 131,746 (30 days), 464,781 (90 days), 1,611,335 (365 days)
install-on-request: 36,378 (30 days), 132,380 (90 days), 478,867 (365 days)
build-error: 0 (30 days)

```

**Reply by carlocab** [reply]
This is the part that tells you it's OpenJDK 17:
```
This is a beta version of openjdk for Apple Silicon
(openjdk 17 preview).
```
Could maybe made to be a bit more prominent. Fixing the version information more fully requires some non-trivial changes to `brew`, which I don't think is worth it since this is meant to be temporary.

**Reply by thesurenk** [reply]
Thanks for the help. But cannot agree.. it maybe an issue for Homebrew to update but this is a big problem for developers, especially when working on commercial software. I have to rollback my brew install and instruct my dev teams not to use "brew install openjdk" on their M1 based  work laptops until this is fixed.

**Reply by SMillerDev** [reply]
> and instruct my dev teams not to use "brew install openjdk" on their M1 based work laptops until this is fixed.

You should always have had that instruction in place for a native install of the OpenJDK. It's on brew to install software (and people would rather have a beta version than no version) and it's on you to see if the latest software for the platform you develop on meets your requirements. Seeing how the only way to make OpenJDK work right now is a pre-release of v17 we will be shipping that and warning users about it's state.

**Reply by carlocab** [reply]
>But cannot agree..

Not 100% clear on what you don't agree with. However, you are welcome to open pull requests to implement the changes you would like to see.

**Reply by thesurenk** [reply]
Showing the correct info on the version being installed would help.

Instead of this referring to "stable 16.0.1" mentioning it is as actually "openjdk 17 preview" and also file/folder named appropriately. That will avoid all the confusion.

```
$brew info openjdk
openjdk: stable 16.0.1 (bottled) [keg-only]
Development kit for the Java programming language
https://openjdk.java.net/
/opt/homebrew/Cellar/openjdk/16.0.1 (618 files, 281.4MB)
  Poured from bottle on 2021-05-31 at 23:05:51
```

**Reply by nandahkrishna** [reply]
As [mentioned before in this thread](https://github.com/Homebrew/discussions/discussions/1594#discussioncomment-811919), mentioning different versions for each architecture requires non-trivial changes to Homebrew. See https://github.com/Homebrew/homebrew-core/pull/69029. Please open pull requests to fix this if you would like to see this behaviour changed.

**Reply by carlocab** [reply]
Another alternative that doesn't require making changes to `brew` is to introduce an `openjdk@17-preview` formula (or some other appropriate name), and fixing the 184 formulae that depend on `openjdk` to depend on `openjdk@17-preview` on ARM, but `openjdk` on Intel.

And then switching it all back once `openjdk` works on ARM.

This will require a significant amount of CI time, and almost undoubtedly a significant amount of time diagnosing and fixing the CI failures that will result from such an undertaking.

If you're willing to work on this, I wouldn't be opposed to it. (I don't speak for other maintainers here, however.)

**Reply by SMillerDev** [reply]
I don't think that's really a solution. A non-versioned formula in brew will always be the latest version available from upstream. In this case that is 17 for ARM (as a preview) and 16 on x86. The only issue here is that brew does not know how to display 2 versions at the same time, which ksurendra can open a PR for if they want.

One further thing to note is that this has been the case since january 15: https://github.com/Homebrew/homebrew-core/pull/68946 when ARM was on Java 16 and x86 on Java 15.

**Reply by adrian-skybaker** [reply]
With all that said, doesn't this part of the question still apply:

>How can I install openjdk 16 ?

Especially now the latest version is 17 even for x86?

**Reply by Bo98** [reply]
[OpenJDK 16 is no longer supported upstream](https://endoflife.date/java). OpenJDK 8, 11 and 17 are the only versions still receiving updates.

We match upstream's support status by only offering those versions (as per our [versioned formulae policy](https://docs.brew.sh/Versions)). If you really need OpenJDK 16, you could use `brew extract` to make a copy of the formula available in your own tap, which you can maintain for whatever you need.
